oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,411 @@
+
+- For example if you are adding a boolean to flag that a learner has not yet activated their account, the default value should be set to False, not None.
+
+The database layer should preserve uniqueness
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- If a model needs to preserve uniqueness between many fields use :code:`unique_together`_.
+.. _unique_together: https://docs.djangoproject.com/en/2.2/ref/models/options/#unique-together
+
+
+
+Best Practices
+--------------
+These practices are designed to help teams create rich and efficient data models within the OpenedX ecosystem.
+They are not standards but guidelines to help teams think about how they are storing data.
+
+Columns should store only exactly what the name describes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- The name of a column in a table should accurately describe the data in that table.
+
+- If a column is named course_id it should only store the course_id. Not the course_key, not a numeric value, not a timestamp. Just the course_id. Conversely if a column is named course_run_key it should store the course run key, not the course_id.
+
+
+Deleting data
+~~~~~~~~~~~~~
+
+-  It is better to have a column to mark the record as inactive than to
+      remove the data from the system using the SQL delete keyword. These models should use Django's
+      :code:`SoftDeletableModel`_.
+.. _SoftDeletableModel: https://django-model-utils.readthedocs.io/en/latest/models.html#softdeletablemodel
+
+
+-  Please note that GDPR may require that data be deleted. If a field
+         is determined to contain PII and falls under the realm of GDPR,
+         That data should be deleted from the system, or obfuscated from the system. `For more information about GDPR and how to delete user data from edx please refer to this documentation`_.
+.. _For more information about GDPR and how to delete user data from edx please refer to this documentation: https://openedx.atlassian.net/wiki/spaces/PLAT/pages/930021733/User+Retirement+Tutorial+for+Developers
+
+
+Don’t trap the data
+~~~~~~~~~~~~~~~~~~~
+
+-  Each piece of information should have its own column. Avoid storing
+      data in blob fields or as JSON in the database.
+
+- Another example is a concatenated string with a separator. It is best to treat these data items in 2 distinct fields.
+
+-  Don’t store encoded (pickle, json, other) objects in the database! If
+      you need to run the python environment to decode the data, analysts who use SQL will have a difficult time querying and decoding this data.
+
+Store everything
+~~~~~~~~~~~~~~~~
+
+-  Storage is cheap!
+
+-  If you are unsure whether you should store something in the database or add history the answer is almost always yes. Store the data and add history. It can be removed later if it is found unnecessary.
```

Comment:
Can we implement a metric for when data can be labeled unnecessary and deleted? Can we implement a process or policy for moving historical data between "hot" and "warm" and "cold" storage?  eg. RDS, S3, Glacier.